Lesson3

profitsw2000 opened the pull request.

profitsw2000 committed:
Создание переменных класса Button и TextView. Добавлена функция, в которой происходит получение объекта кнопок и текстового поля.

profitsw2000 committed:
Добавлен вызов функции findViews().

profitsw2000 committed:
Создание метода buttonsListen(), вызов метода в onCreate().

profitsw2000 committed:
Создание метода setListener(), в качестве параметра ему передаётся объект кнопки, которую необходимо будет "слушать". Создан класс Calculator для основных операций калькулятора.

profitsw2000 committed:
Созданы методы getButton() и setText() класса Calculator.

profitsw2000 committed:
Создан нумерованный список и метод equalPressed() в классе Calculator.

profitsw2000 committed:
Переделаны функции getButton() и equalPressed(), в них исправлены ошибки в логике работы.

profitsw2000 committed:
Добавлены методы onSaveInstanceState() и onRestoreInstanceState() для сохранения введённых пользователем данных.

profitsw2000 committed:
В классе Calculator удалены два метода, некорректно работавших при выводе результатов в текстовое поле, написан новый метод, позволяющий правильно отображать результат работы калькулятора.

profitsw2000 committed:
Создан макет калькулятора для горизонтальной ориентации эрана, подключен к activity_main.

profitsw2000 committed:
Добавил код для корректной работы:
-при введении нескольких нулей перед значащей цифрой;
-при нажатии запятой перед вводом числа (приложение вылетало).
Добавлено отображение первого аргумента на второй строке при нажатии кнопки операции вычисления (+,- и т.д.)

profitsw2000 committed:
Изменен параметр VerticalBias для элемента TextView.

DwarfPorter commented on app/src/main/java/ru/profitsw2000/calculatorsw/MainActivity.java:
```diff
@@ -1,14 +1,98 @@
 public class MainActivity extends AppCompatActivity {
 
+    private final static String keyCalculator = "AAAAA"    ;
+    private Button one, two, three, four, five, six, seven, eight, nine, zero,
```
Обычно в командах каждую переменную объявляют на новой строке, так в дальнейшем легче менять код, проще выкинуть переменную, которая уже не используется

DwarfPorter commented on app/src/main/java/ru/profitsw2000/calculatorsw/MainActivity.java:
```diff
@@ -1,14 +1,98 @@
+    @Override
+    protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
+        super.onRestoreInstanceState(savedInstanceState);
+        Calculator calculator = new Calculator()    ;
```
Не используется, выкинуть

DwarfPorter commented on app/src/main/java/ru/profitsw2000/calculatorsw/Calculator.java:
```diff
@@ -0,0 +1,318 @@
+                var2 = Double.parseDouble(argument2.toString()) ;
+                //add second line
+                outputText = convertString(argument1.toString()) + convertString(argument2.toString())   ;
+            }
```
Мне показалось, что тут есть дублирование кода. В обоих ветках отличается argument1 и argument2, var1 и var2. Может сделать как приватный метод?
```
private Double parseFromArgument(argument){
    if (argument.equals(0)){
    ....

}

```

DwarfPorter commented on app/src/main/java/ru/profitsw2000/calculatorsw/Calculator.java:
```diff
@@ -0,0 +1,318 @@
+        else {
+            switch (buttonText) {
+                case ",":
+                    if (this.operation == Operation.NONE) {
```
Каждый case  лучше оформить как отдельный метод, будет лучше читаться код

DwarfPorter commented on app/src/main/java/ru/profitsw2000/calculatorsw/Calculator.java:
```diff
@@ -0,0 +1,318 @@
+                    this.operation = Operation.NONE ;
+                    outputText = "0"   ;
+                    break;
+                case "<--":
```
все подобные "<--", "C" и т.п. привести к константам

DwarfPorter replied:
А может быть тоже их как enum сделать?

DwarfPorter reviewed (approved):
В целом неплохо, но лучше над кодом еще поработать.

DwarfPorter commented on app/src/main/java/ru/profitsw2000/calculatorsw/Calculator.java:
```diff
@@ -0,0 +1,318 @@
+        this.operation = Operation.NONE ;
+    }
+
+    public void setText(TextView textView) {
```
Не надо ссылаться на TextView из этого класса, он ничего знать не должен о TextView

profitsw2000 committed:
Доработан код в соответствии с рекомендациями преподвателя:
-каждая переменная объявлена в новой строке;
-удален код создания нового объекта класса Calculator в методе onRestoreInstanceState(@NonNull Bundle savedInstanceState);
-устранено дублирование кода - создан метод parseFromArgument();
-каждый case оформлен как отдельный метод;
-строки типа "C", "<--" и т.п. приведены к константам;
-метод, ссылающийся на TextView из класса Calculator удалён, передача строки в TextView осуществлено другим способом.
